[Merged by Bors] - feat(algebra/operations): add three lemmas

faenuccio opened the pull request:
Add lemmas `one_le_inv`, `self_le_self_inv` and `self_inv_le_one`

---
<!--
put comments you want to keep out of the PR commit here.
If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

faenuccio committed:
chore(algebra/operations): add three lemmas

bryangingechen commented on src/algebra/algebra/operations.lean:
```diff
@@ -287,6 +288,29 @@ lemma le_div_iff {I J K : submodule R A} : I ≤ J / K ↔ ∀ (x ∈ I) (z ∈
+(1 : submodule R A) ≤ (1 : submodule R A) / I :=
+begin
+  rw [le_div_iff_mul_le, one_mul], assumption,
+end
```
```suggestion
by rwa [le_div_iff_mul_le, one_mul]
```

faenuccio committed:
Update src/algebra/algebra/operations.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

eric-wieser commented on src/algebra/algebra/operations.lean:
```diff
@@ -4,6 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Kenny Lau
 -/
 import algebra.algebra.basic
+import linear_algebra.finsupp
```
Is this needed? I don't see an obvious mention of finsupp below

faenuccio replied:
You're right: it is not (it was in a previous implementation and forgot it there), thanks.

faenuccio committed:
Update operations.lean

eric-wieser commented on src/algebra/algebra/operations.lean:
```diff
@@ -287,6 +287,27 @@ lemma le_div_iff {I J K : submodule R A} : I ≤ J / K ↔ ∀ (x ∈ I) (z ∈
 by rw [le_div_iff, mul_le]
 
+lemma one_le_inv {I : submodule R A} (hI : I ≤ (1 : submodule R A)) :
+(1 : submodule R A) ≤ (1 : submodule R A) / I :=
```
```suggestion
  (1 : submodule R A) ≤ (1 : submodule R A) / I :=
```

faenuccio committed:
Update src/algebra/algebra/operations.lean

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

kbuzzard commented on src/algebra/algebra/operations.lean:
```diff
@@ -287,6 +287,27 @@ lemma le_div_iff {I J K : submodule R A} : I ≤ J / K ↔ ∀ (x ∈ I) (z ∈
 by rw [le_div_iff, mul_le]
 
+lemma one_le_inv {I : submodule R A} (hI : I ≤ (1 : submodule R A)) :
+  (1 : submodule R A) ≤ (1 : submodule R A) / I :=
```
```suggestion
lemma one_le_inv {I : submodule R A} (hI : I ≤ 1) :
  (1 : submodule R A) ≤ 1 / I :=
```

kbuzzard replied:
Lean reads from left to right : by the time it knows you're talking about inequalities of submodules (by reading the LHS) it assumes that the RHS has the same type.

faenuccio replied:
Ah, thanks, good to keep in mind.

kbuzzard commented on src/algebra/algebra/operations.lean:
```diff
@@ -287,6 +287,27 @@ lemma le_div_iff {I J K : submodule R A} : I ≤ J / K ↔ ∀ (x ∈ I) (z ∈
+  (1 : submodule R A) ≤ (1 : submodule R A) / I :=
+by rwa [le_div_iff_mul_le, one_mul]
+
+lemma self_le_self_inv {I : submodule R A} (hI : I ≤ (1 : submodule R A)) :
```
```suggestion
lemma self_le_self_inv {I : submodule R A} (hI : I ≤ 1) :
```

faenuccio committed:
Update src/algebra/algebra/operations.lean

Co-authored-by: Kevin Buzzard <k.buzzard@imperial.ac.uk>

faenuccio committed:
Update src/algebra/algebra/operations.lean

Co-authored-by: Kevin Buzzard <k.buzzard@imperial.ac.uk>

faenuccio committed:
change inv to div and add one iff

jcommelin commented on src/algebra/algebra/operations.lean:
```diff
@@ -287,6 +287,30 @@ lemma le_div_iff {I J K : submodule R A} : I ≤ J / K ↔ ∀ (x ∈ I) (z ∈
 by rw [le_div_iff, mul_le]
 
+lemma one_le_div {I : submodule R A} :
+  I ≤ (1 : submodule R A) ↔ 1 ≤ 1 / I :=
```
```suggestion
@[simp] lemma one_le_one_div {I : submodule R A} :
  1 ≤ 1 / I ↔ I ≤ (1 : submodule R A) :=
```
(You'll need to swap the subproofs as well.)

jcommelin commented on src/algebra/algebra/operations.lean:
```diff
@@ -287,6 +287,30 @@ lemma le_div_iff {I J K : submodule R A} : I ≤ J / K ↔ ∀ (x ∈ I) (z ∈
+  {rw [le_div_iff_mul_le, one_mul] at hI, assumption },
+end
+
+lemma self_le_self_div {I : submodule R A} (hI : I ≤ 1) :
```
```suggestion
lemma le_self_mul_one_div {I : submodule R A} (hI : I ≤ 1) :
```

jcommelin commented on src/algebra/algebra/operations.lean:
```diff
@@ -287,6 +287,30 @@ lemma le_div_iff {I J K : submodule R A} : I ≤ J / K ↔ ∀ (x ∈ I) (z ∈
+  apply mul_le_mul_right (one_le_div.mp hI),
+end
+
+lemma self_div_le_one {I : submodule R A} : I * (1 / I) ≤ 1 :=
```
```suggestion
lemma mul_one_div_le_one {I : submodule R A} : I * (1 / I) ≤ 1 :=
```

jcommelin commented on src/algebra/algebra/operations.lean:
```diff
@@ -287,6 +287,30 @@ lemma le_div_iff {I J K : submodule R A} : I ≤ J / K ↔ ∀ (x ∈ I) (z ∈
+  I ≤ (1 : submodule R A) ↔ 1 ≤ 1 / I :=
+begin
+  split, all_goals {intro hI},
+  {rw [le_div_iff_mul_le, one_mul], assumption },
```
Seems like you do this `rw` in both subproofs, so you could do it before the `split` instead.
(Also, note that `rwa [..]` is short for `rw [..], assumption`.)

faenuccio replied:
Thanks for the `rwa[...]`; on the other hand, the `rw` is not `at` the same point, I am not sure how to perform it before `split`.

faenuccio committed:
Update src/algebra/algebra/operations.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

faenuccio committed:
Update src/algebra/algebra/operations.lean

Can I ask why the inversion on line 291? I am of course accepting it, but am not sure to understand the reason. Also, I guess that the `( 1 : submodule R A)` will now appear the first term.

Co-authored-by: Johan Commelin <johan@commelin.net>

faenuccio committed:
implement

faenuccio committed:
Update src/algebra/algebra/operations.lean

Co-authored-by: Johan Commelin <johan@commelin.net>

faenuccio committed:
remove two spaces before }

jcommelin reviewed (approved):
Thanks :tada:

bors merge
